File: problem_sets/w1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long num;
    do
    {
        num = get_long("Number: ");
    }
    while (num < 1);

    int count = 0;
    int digitSum = 0;
    int nonMultipSum = 0;
    int firstTwoDig;
    int firstDig;

    // for (long i = num; i > 0; i = i / 10)
    for (; num > 0; num /= 10)
    {
        count++;
        int digit = num % 10;

        // sum of the digits not multiplied by 2
        if (count % 2 != 0)
        {
            nonMultipSum += digit;
        }
        else
        {
            // multiply every other digit by 2
            int product = digit * 2;

            // add the products' digits together
            digitSum += (product / 10 + product % 10);
        }

        if (num > 9 && num < 100)
        {
            firstTwoDig = num;
        }
        else if (num < 10)
        {
            firstDig = num;
        }
    }

    int final = digitSum + nonMultipSum;
    bool valid = final % 10 == 0;

    if (valid)
    {
        if (count == 15 && (firstTwoDig == 34 || firstTwoDig == 37))
        {
            printf("AMEX\n");
        }
        else if (count == 16 && (firstTwoDig == 51 || firstTwoDig == 52 || firstTwoDig == 53 ||
                                 firstTwoDig == 54 || firstTwoDig == 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((count == 13 || count == 16) && firstDig == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

File: problem_sets/w6/sentimental-credit/credit.py
from cs50 import get_string, get_int
import re


def main():
    # Prompt the user for the card number
    number = get_int("Number: ")
    number = str(number)
    numberLen = len(number)

    valid = validateCardNumber(number, numberLen)
    detectCardNetwork(number, valid)


# Check whether the card number is valid
def validateCardNumber(number, numberLen):

    # Multiply every other digit by 2, starting with the number’s second-to-last digit,
    # and then add those products’ digits together
    digitProductSum = 0
    for i in range(numberLen - 2, -1, -2):
        multipliedDigit = int(number[i]) * 2
        digitProductSum += multipliedDigit // 10 + multipliedDigit % 10

    # Add the sum to the sum of the digits that weren’t multiplied by 2
    for i in range(numberLen - 1, -1, -2):
        digitProductSum += int(number[i])

    # If the total’s last digit is 0, the number is valid
    if digitProductSum % 10 == 0:
        return (True)
    else:
        return (False)


# If the card number is valid, detect the card network, with regex
def detectCardNetwork(number, valid):
    if valid:
        if re.match("^3[4|7][0-9]{13}", number):
            print("AMEX")
        elif re.match("5[1-5][0-9]{14}", number):
            print("MASTERCARD")
        elif re.match("4[[0-9]{12}|[0-9]{15}]", number):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


main()

# If the card number is valid, detect the card network, without the regex
# def detectCardNetwork(number, numberLen, valid):
#    if valid:
#        if numberLen == 15 and (number[0:2] in ["34", "37"]):
#            print("AMEX")
#        elif numberLen == 16 and (number[0:2] in ["51", "52", "53", "54", "55"]):
#            print("MASTERCARD")
#        elif numberLen in [13, 16] and number[0] == "4":
#            print("VISA")
#        else:
#            print("INVALID")
#    else:
#        print("INVALID")
